src/Plutarch/Test/Export.hs

```diff
@@ -0,0 +1,728 @@
+
+ @since 1.2.1
+-}
+newtype WithExport (dat :: Type) (red :: Type) (info :: Type) (a :: Type)
+  = WithExport (RWS (WithExportEnv dat red info) WithExportLog () a)
+  deriving
+    ( -- | @since 1.2.1
+      Functor
+    , -- | @since 1.2.1
+      Applicative
+    , -- | @since 1.2.1
+      Monad
+    )
+    via ( RWS (WithExportEnv dat red info) WithExportLog ()
+        )
+
+{- | Given a name for the overall test suite, a raw export, a linker, and some
+ test parameters, assemble a 'TestTree'.
+
+ This acts as a 'testGroup' replacement, and can be used in a similar way. The
+ exact number of tests this will generate depends on the script export given,
+ and the scripts we ask to test:
+
+ - If linking fails, you will get a single, always-failing, test, which
+ presents the linker error;
+ - If any tests request non-existent scripts (that is, a name that doesn't
+ correspond to an existing script), those requests will be bundled into a
+ single, always-failing, test, which presents the name that doesn't have a
+ script associated with it;
+ - Otherwise, one test per request.
+
+ @since 1.2.1
+-}
+exportTests ::
+  forall (dat :: Type) (red :: Type) (info :: Type) (param :: Type).
+  String ->
+  RawScriptExport ->
+  Linker param (ScriptExport info) ->
+  TestParams dat red param ->
+  WithExport dat red info () ->
+  TestTree
+exportTests name raw linker args comp = case runWithExport comp raw linker args of
+  Left err -> singleTest "Linking" . FailedToLink $ err
+  Right result -> testGroup name . toList . Map.foldMapWithKey go $ result
+    where
+      go :: Text -> PreparedScriptTests -> Acc TestTree
+      go testName = \case
+        NoSuchScript -> pure . singleTest "Presence" . MissingScript $ testName
+        PreparedTests tests -> fmap prepDataToTest tests
+
+{- | States that the named validator should have the given outcome, using the
+ currently in-scope arguments. The test will be labelled with the given
+ explanation (or reason) in the suite, as long as the named script actually
+ exists.
+
+ = Example
+
+ > tests = do
+ >    assertValidator Crashes "reason for crash" "myScript"
+ >    assertValidator Runs "reason for success" "aBetterScript"
+
+ @since 1.2.1
+-}
+assertValidator ::
+  forall (dat :: Type) (red :: Type) (info :: Type).
+  (ToData dat, ToData red) =>
+  -- | What you expect to happen
+  ScriptOutcome ->
+  -- | The reason or explanation to label the test with
+  String ->
+  -- | The name of the script to run
+  Text ->
+  WithExport dat red info ()
+assertValidator outcome reason name = WithExport $ do
+  lookedUp <- asks (view (#raws % #rawScripts % at' @(Map _ _) name))
+  case lookedUp of
+    Nothing -> tell . noSuchScript $ name
+    Just envelope -> case view (to tsRole) envelope of
+      ValidatorRole -> do
+        datum' <- asks (toData . view #datum)
+        redeemer' <- asks (toData . view #redeemer)
+        sc' <- asks (toData . view #context)
+        let script = view (to tsScript) envelope
```

Comment:
CRITICAL!!!!

This should not use raw script. We need to look up linked script. This will not work if we use parameterized script(unlinked script).

Reply:
We need to look up twice one for rawScripts and other for exportedScripts. 
This will add new type of error that will say "script only exist on {export|raw}".

`WithExportEnv` doesn't hold `ScriptExport` so that needs to be added as well.

Reply:
Ah, OK, that makes sense. I thought this seemed very odd, so it's good to know I was wrong there.

So does this mean the environment needs to hold _both_ the raw (so we can establish the role) _and_ the linked script collections?